File: frontend/src/api/games.ts
import { http } from "./http"

export interface Game {
  id: number
  game_date: string
  home_team: string
  away_team: string
}

export async function fetchGames(): Promise<Game[]> {
  const { data } = await http.get<Game[]>("/games")
  return data
}

export interface Event {
  id: number
  game_date: string
  home_team: string
  away_team: string
  period: number
  clock: string
  home_team_skaters: number
  away_team_skaters: number
  team: string
  player: string
  event: string
  detail_1?: string | null
  detail_2?: string | null
  detail_3?: string | null
  detail_4?: string | null
  x_coordinate?: number | null
  y_coordinate?: number | null
  player_2?: string | null
  x_coordinate_2?: number | null
  y_coordinate_2?: number | null
}

export async function fetchGameEvents(gameId: number): Promise<Event[]> {
  const { data } = await http.get<Event[]>(`/games/${gameId}/events`)
  return data
}

export interface ShotCoordinate {
  x: number
  y: number
}

export async function fetchGameShotDensity(gameId: number): Promise<ShotCoordinate[]> {
  const { data } = await http.get<ShotCoordinate[]>(`/games/${gameId}/shot-density`)
  return data
}

export async function fetchGameGoalDensity(gameId: number): Promise<ShotCoordinate[]> {
  const { data } = await http.get<ShotCoordinate[]>(`/games/${gameId}/goal-density`)
  return data
}

export async function fetchEventTypes(): Promise<string[]> {
  const { data } = await http.get<string[]>("/events/types")
  return data
}

export interface PlayerInfo {
  id: number
  name: string
  number?: number | null
}

export async function fetchPlayers(): Promise<PlayerInfo[]> {
  const { data } = await http.get<PlayerInfo[]>("/players")
  return data
}

export interface GameExport {
  game: {
    id: number
    game_date: string
    home_team: string
    away_team: string
  }
  events: Event[]
}

export async function exportGameData(gameId: number): Promise<GameExport> {
  const { data } = await http.get<GameExport>(`/games/${gameId}/export`)
  return data
}

File: backend/src/routes/games.py
"""Routes related to games and game-derived data."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import distinct, select

from ..db.database import get_db
from ..models import Game, Event
from ..schemas import GameSchema, EventSchema
from ..schemas.shot import ShotCoordinateSchema

router = APIRouter(prefix="/games", tags=["Games"])


@router.get("", response_model=list[GameSchema])
async def list_games(db: Session = Depends(get_db)):
    """Return all *Game* records."""
    return db.query(Game).all()


@router.get("/{game_id}/events", response_model=list[EventSchema])
async def game_events(game_id: int, db: Session = Depends(get_db)):
    """Return all events belonging to the given game id."""
    game_obj = db.query(Game).filter(Game.id == game_id).first()
    if not game_obj:
        raise HTTPException(status_code=404, detail="Game not found")

    events = (
        db.query(Event)
        .filter(
            Event.game_date == game_obj.game_date,
            Event.home_team == game_obj.home_team,
            Event.away_team == game_obj.away_team,
        )
        .all()
    )
    return events


@router.get("/{game_id}/shot-density", response_model=list[ShotCoordinateSchema])
async def game_shot_density(game_id: int, team: str | None = None, db: Session = Depends(get_db)):
    """Return all (x, y) shot coordinates for the given game.

    Optionally filter by *team* (home/away/team name).
    """
    game_obj = db.query(Game).filter(Game.id == game_id).first()
    if not game_obj:
        raise HTTPException(status_code=404, detail="Game not found")

    query = (
        db.query(Event.x_coordinate.label("x"), Event.y_coordinate.label("y"))
        .filter(
            Event.game_date == game_obj.game_date,
            Event.home_team == game_obj.home_team,
            Event.away_team == game_obj.away_team,
            Event.event.ilike("shot"),
            Event.x_coordinate.isnot(None),
            Event.y_coordinate.isnot(None),
        )
    )

    if team:
        query = query.filter(Event.team == team)

    rows = query.all()
    return [{"x": r.x, "y": r.y} for r in rows]


@router.get("/{game_id}/goal-density", response_model=list[ShotCoordinateSchema])
async def game_goal_density(game_id: int, team: str | None = None, db: Session = Depends(get_db)):
    """Return all (x, y) goal coordinates for the given game."""
    game_obj = db.query(Game).filter(Game.id == game_id).first()
    if not game_obj:
        raise HTTPException(status_code=404, detail="Game not found")

    query = (
        db.query(Event.x_coordinate.label("x"), Event.y_coordinate.label("y"))
        .filter(
            Event.game_date == game_obj.game_date,
            Event.home_team == game_obj.home_team,
            Event.away_team == game_obj.away_team,
            Event.event.ilike("goal"),
            Event.x_coordinate.isnot(None),
            Event.y_coordinate.isnot(None),
        )
    )
    if team:
        query = query.filter(Event.team == team)

    rows = query.all()
    return [{"x": r.x, "y": r.y} for r in rows]


@router.get("/{game_id}/export")
async def export_game_data(game_id: int, db: Session = Depends(get_db)):
    """Return full game metadata and all associated events for the given game id."""
    from fastapi.responses import JSONResponse  # imported here to avoid circular imports

    # Fetch game
    game_obj = db.query(Game).filter(Game.id == game_id).first()
    if not game_obj:
        raise HTTPException(status_code=404, detail="Game not found")

    events_q = (
        db.query(Event)
        .filter(
            Event.game_date == game_obj.game_date,
            Event.home_team == game_obj.home_team,
            Event.away_team == game_obj.away_team,
        )
        .all()
    )

    game_data = GameSchema.model_validate(game_obj).model_dump()
    event_data = [EventSchema.model_validate(ev).model_dump() for ev in events_q]

    payload = {"game": game_data, "events": event_data}

    safe_home = game_data["home_team"].replace(" ", "_")
    safe_away = game_data["away_team"].replace(" ", "_")
    filename = f"{game_data['game_date']}_{safe_home}_vs_{safe_away}.json"

    return JSONResponse(content=payload, headers={"Content-Disposition": f"attachment; filename=\"{filename}\""})
